doctests

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+
+"""
+A shell-executable script to run all the commands contained in README.md or specified in EXTRA_TESTS and
+assert that Wheatley does not crash.  With decent test coverage, this prevents mistakes like silly crashes
+and argument errors.
+"""
+
+from subprocess import STDOUT, check_output, CalledProcessError, TimeoutExpired
+import shlex
+
+# ID of a tower that I made called 'DO NOT ENTER'
+ROOM_ID = "238915467"
+EXAMPLE_METHOD = "Plain Bob Major"
+IGNORE_STRING = "<!--- doctest-ignore -->"
+
+
+# Tests that should be run on top of the examples from README.md
+EXTRA_TESTS = []
+
+
+def get_all_tests():
+    """ Get all commands that should be tested (these should start with 'wheatley [ID NUMBER]'). """
+
+    tests = []
+
+    with open("README.md") as readme:
+        is_in_code = False
+        ignore_next_examples = False
+
+        for i, l in enumerate(readme.read().split("\n")):
+            stripped_line = l.strip()
+
+            # If we see the line IGNORE_STRING, we should ignore the next block of examples
+            if not is_in_code and stripped_line == IGNORE_STRING:
+                ignore_next_examples = True
+
+            if stripped_line.startswith("```"):
+                # Clear the ignore flag whenever we *finish* a code block
+                if is_in_code:
+                    ignore_next_examples = False
+
+                # Alternate into or out of code blocks
+                is_in_code = not is_in_code
+
+            # If this is a code line that starts with `wheatley`, then it should be used as an example
+            if is_in_code and stripped_line.startswith("wheatley"):
+                if ignore_next_examples:
+                    print(f"Ignoring README.md:{i}: {stripped_line}")
+                else:
+                    tests.append((f"README.md:{i}", stripped_line))
+
+    return tests + [(f"EXTRA_TESTS:{i}", cmd) for i, cmd in EXTRA_TESTS]
+
+
+def command_to_converted_args(location, command):
+    # Bit of a hack, but this will convert commands starting with 'wheatley' to start with './run-wheatley'
+    # and make sure that [ID NUMBER] is replaced with a valid room ID.  This way, the tests will run on the
+    # version contained in the current commit, rather than a release version.
+    edited_command = "./run-" + command.replace("[ID NUMBER]", ROOM_ID) \
+                                       .replace("[METHOD TITLE]", '"' + EXAMPLE_METHOD + '"')
+
+    args = shlex.split(edited_command)
+    args.insert(1, "integration-test")
+
+    return (args, location, edited_command)
+
+
+
+# I'm not sure how to do this in python.  I want `run_test` to return one of 3 types: an 'OK', a 'TIMEOUT'
+# or an 'ERROR' with a return code and an output.  These two classes and None are trying to represent the
+# enum:
+#
+# enum TestResult {
+#     Ok,
+#     Timeout,
+#     Error(usize, String),
+# }
```

Comment:
I think this way is fine, I don't know a more idiomatic approach atm.

Reply:
Ah, OK.  I'm not sure whether or not to remove the comment, on the basis that it does a pretty good job of describing what the code is trying to achieve.

Reply:
I think it should still be there.